### DISCUSSION TITLE
Is Go 1.17 required to build from source?

### DISCUSSION THREAD

**Discussion by srid** [zk-org/zk#76] [Help]
If so, perhaps the README should say so. On NixOS,

```
❯ nix-shell -p go -p icu --run make
CGO_ENABLED=1 go build -tags "fts5 icu" -ldflags "-X=main.Version=`git describe --tags --match v[0-9]* 2> /dev/null` -X=main.Build=`git rev-parse --short HEAD`" 
# github.com/tliron/kutil/logging/simple
../../go/pkg/mod/github.com/tliron/kutil@v0.1.48/logging/simple/backend.go:79:17: undefined: io.Discard
note: module requires Go 1.17
make: *** [Makefile:3: build] Error 2
```

**Comment by srid** [comment]
As an aside, it would be nice if there was either a Docker image or a static binary (Linux) available for `zk` - so that I can try it out on my Zettelkasten locally. Been wondering about this project for a while, to see if there is something that can be used as inspiration for improving neuron/emanote ...

**Comment by mickael-menu** [comment]
> If so, perhaps the README should say so.

`zk` itself depends on Go 1.15, but some dependencies might depend on later versions. Since I don't want to monitor the version of each dependency and the error is clear enough, I won't mention it in the README.

> As an aside, it would be nice if there was either a Docker image or a static binary (Linux) available for zk - so that I can try it out on my Zettelkasten locally.

You can find static binaries for each release on the [Releases page](https://github.com/mickael-menu/zk/releases).

A Docker image would be really nice indeed. [There has been some discussion about it here](https://github.com/mickael-menu/zk/issues/53#issuecomment-864466160), but I didn't take the time to figure out how to adapt `zk` for the limited FS scope available through Docker.

**Reply by srid** [reply]
> You can find static binaries for each release on the Releases page.

Those are actually dynamically linked binaries.

```
$ ldd zk
	linux-vdso.so.1 (0x00007fffb9d89000)
	libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007fdfb66e4000)
	libm.so.6 => /usr/lib64/haswell/libm.so.6 (0x00007fdfb6602000)
	libicuuc.so.63 => not found
	libicui18n.so.63 => not found
	libdl.so.2 => /usr/lib64/libdl.so.2 (0x00007fdfb65fd000)
	libc.so.6 => /usr/lib64/haswell/libc.so.6 (0x00007fdfb63c1000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fdfb6700000)
```

**Reply by mickael-menu** [reply]
Ha I missed that, looks like importing C libraries in Go disable static linking by default. I'll take a look into that. Thanks for spotting this issue.

**Reply by mickael-menu** [reply]
I don't think I'll be able to provide static binaries for `zk`, it's quite messy:

* A critical dependency is using glibc which doesn't really work with static linking.
* It's not supported on macOS with CGO.

Docker should be a better option/alternative if I figure it out.
